extra/liquidexport

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+"""liquidexport – host a metrics HTTP endpoint with Prometheus formatted data from liquidctl
+
+This is an experimental script that collects stats from liquidctl and exposes them as a http://localhost:8098/metrics
+endpoint in the Prometheus text format.
+See: https://prometheus.io/docs/instrumenting/exposition_formats/#text-format-example
+
+Example metric with labels:
+# HELP liquidctl liquidctl exported metrics
+# TYPE liquidctl gauge
+liquidctl{device="NZXT Kraken X (X42, X52, X62 or X72)",sensor="liquid_temperature",unit="°C"} 33.6
+
+Usage:
+  liquidexport [options]
+  liquidexport --help
+  liquidexport --version
+
+Options:
+  --interval <seconds>    Update interval in seconds [default: 2]
+  --hid <module>          Override API for USB HIDs: usb, hid or hidraw
+  --legacy-690lc          Use Asetek 690LC in legacy mode (old Krakens)
+  --match <name>          Filter devices by name
+  --vendor <id>           Filter devices by vendor id
+  --product <id>          Filter devices by product id
+  --release <number>      Filter devices by release number
+  --serial <number>       Filter devices by serial number
+  --bus <bus>             Filter devices by bus
+  --address <address>     Filter devices by address in bus
+  --usb-port <port>       Filter devices by USB port in bus
+  --pick <number>         Pick among many results for a given filter
+  --port <number>         Port for the HTTP /metrics endpoint
+  -v, --verbose           Output additional information
+  -g, --debug             Show debug information on stderr
+  --version               Display the version number
+  --help                  Show this message
+
+Examples:
+  liquidexport
+  liquidexport --product 0xb200
+  liquidexport --interval 0.5
+
+Copyright (C) 2020  Alex Berryman
+SPDX-License-Identifier: GPL-3.0-or-later
+"""
+
+import logging
+import sys
+import time
+
+import liquidctl.cli as _borrow
+import usb
+
+from docopt import docopt
+from liquidctl.driver import *
+
+from prometheus_client import start_http_server
+from prometheus_client.core import GaugeMetricFamily, REGISTRY
+
+import re
+
+LOGGER = logging.getLogger(__name__)
+
+
+def gauge_name_sanitize(name):
+    name = re.sub(r'(?<!^)(?=[A-Z])', '_', name).lower()
+    name = name.replace(" ", "_")
+    return name
+
+
+class LiquidCollector(object):
+    def __init__(self, args):
+        self.description = 'liquidctl exported metrics'
+        self.args = args
+
+    def collect(self):
+        labels = ['device', 'sensor', 'unit']
+        g = GaugeMetricFamily('liquidctl', self.description, labels=labels)
+        for d in devs:
+            d.connect()
+            try:
+                get_status = d.get_status()
+                for metric in get_status:
+                    # Only allow numerical stats and exclude string values like "firmware='6.3.3'"
+                    if isinstance(metric[1], (int, float, complex)) and not isinstance(metric[1], bool):
```

Comment:
This filter doesn't match `datetime.timedelta`s; we have a couple of those in one driver for PSU uptime (session and ever).

On the other hand, have you considered filtering out all metrics without a unit?

---

If you haven't already seen it, the CLI also implements some [conditional formatting of status values](https://github.com/jonasmalacofilho/liquidctl/blob/de95601ba4cc2d4b59e600d5a691669e6b56e16f/liquidctl/cli.py#L190-L195).

Reply:
I added support for the `datetime.timedelta` values. They will be cast into `seconds` and have their `unit` updated to reflect that before the data is used to create a GaugeMetric. I tested this by faking the following timedelta which yields the following metric:
```python
get_status.append(('Current uptime', timedelta(seconds=42), ''))
```
`liquidctl{device="NZXT Smart Device V2",sensor="current_uptime",unit="seconds"} 42.0`

I also added support for `firmware` and other unit-less (non-timedelta) statues as an InfoMetric:
`liquidctl_info{device="NZXT Kraken X (X42, X52, X62 or X72)",firmware_version="6.0.2"} 1.0`